linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/Emitter.scala

```diff
@@ -411,27 +411,49 @@ final class Emitter(config: Emitter.Config, prePrinter: Emitter.PrePrinter) {
     else lhs.className.compareTo(rhs.className) < 0
   }
 
-  private def genClass(linkedClass: LinkedClass,
+  private def genClass(
+      linkedClass_! : LinkedClass, // scalastyle:ignore
       moduleContext: ModuleContext): GeneratedClass = {
-    val className = linkedClass.className
+
+    /* !!! In this method, *every* time you use linkedClass_!, you must justify
+     * why you have the right to access the fields you are reading.
+     * That's why we give it that dangerous-looking name.
+     */
+
+    // Cache identity; always safe to access
+    val kind = linkedClass_!.kind
+    val isJSClass = kind.isJSClass // we use this one a lot
+    val className = linkedClass_!.className
+    val ancestors = linkedClass_!.ancestors
 
     val classCache = classCaches.getOrElseUpdate(
-        new ClassID(linkedClass.ancestors, moduleContext), new ClassCache)
+        new ClassID(kind, ancestors, moduleContext), new ClassCache)
 
     var changed = false
     def extractChanged[T](x: (T, Boolean)): T = {
       changed ||= x._2
       x._1
     }
 
-    val classTreeCache = extractChanged {
-      classCache.getCache(linkedClass.version,
-          hasInstances = linkedClass.hasInstances,
-          hasRuntimeTypeInfo = linkedClass.hasRuntimeTypeInfo,
-          hasInstanceTests = linkedClass.hasInstanceTests)
-    }
+    /* The class version itself; it's OK to get that one as long as we don't
+     * use it to *produce* trees, which we should never do anyway.
+     */
+    val classVersion = linkedClass_!.version
 
-    val kind = linkedClass.kind
+    val classTreeCache = extractChanged(classCache.getCache(classVersion))
+
+    /* Information that we can use for uncached decision-making.
+     * We call "decision-making" any if/else branch not covered by an explicit
+     * cache.
+     *
+     * Note that it is *not* safe to use uncachedDecisions in *cached*
+     * decision-making!
+     */
+    val uncachedDecisions = extractChanged(classCache.getUncachedDecisions(linkedClass_!))
+
+    // Delegate justifications for those two to their use sites
+    val (linkedInlineableInit_!, linkedMethods_!) =
+      classEmitter.extractInlineableInit(linkedClass_!)(classCache)
```

Comment:
How do we make sure the results of this are only used in places that are guarded by the `classCache`?

IIUC, the results may depend on global knowledge.